File: Python/DynamicProgramming/#5.py
# Given a string s, return the longest palindromic substring in s.
# Input: s = "babad"
# Output: "bab"
# Explanation: "aba" is also a valid answer.

class Solution:
    def longestPalindrome(self, s: str) -> str:
        size = len(s)
        dp = [[False] * size for i in range(size)]
        res = s[0:1]
        
        def helper(i, j):
            nonlocal res
            if s[i] == s[j] and (i - j < 2 or dp[j + 1][i - 1]):
                dp[j][i] = True
                if len(res) < i - j + 1:
                    res = s[j: i + 1]

            return

        for i in range(size):
            dp[i][i] = True
        for i in range(size):
            for j in range(i - 1, -1, -1):
                helper(i, j)
        return res
        

File: Java/DynamicProgramming/#5.java
// Given a string s, return the longest palindromic substring in s.
// Input: s = "babad"
// Output: "bab"
// Explanation: "aba" is also a valid answer.

class Solution {
    
    String res = "";
    int length = 0;
    
    public boolean helper(int dp[][], String s, int i, int j) { 
        if (dp[i][j] != -1) {
            return dp[i][j] == 1? true: false;
        } 
        boolean ans;
        if (i == j) {
            dp[i][j] = 1;
            ans = true;
        }
        else if (i > j) {
            return true;
        }
        else {
            ans = (s.charAt(i) == s.charAt(j) & helper(dp, s, i + 1, j - 1));
        }
        if (ans) {
            dp[i][j] = 1;
            if (length < j - i + 1) {
                res = s.substring(i, j + 1);
                length = j - i + 1;
            }
        }
        else {
            dp[i][j] = 0;
        }
        return ans;
        
    }
    
    public String longestPalindrome(String s) {
        int n = s.length();
        int dp[][] = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                dp[i][j] = -1;
            }
        }
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                helper(dp, s, i, j);
            }
        }
        return res;
        

    }
}
